ios/RCTHTTPRequestHandler+AuthenticationChallenge.m

```diff
@@ -0,0 +1,189 @@
+#import <objc/runtime.h>
+#import <Foundation/Foundation.h>
+#import <React/RCTHTTPRequestHandler.h>
+
+@implementation RCTHTTPRequestHandler (AuthenticationChallengeExtension)
+
+/* The load method is called by the Objective-C runtime when the class is loaded into memory,
+ * it happens early in the application's lifecycle.
+ */
++ (void)load {
+  // dispatch_once avoids the load method to run twice
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    NSLog(@"[RCTHTTPRequestHandler extension] It will swizzle authentication challenge on this data delegate.");
+    Class class = [RCTHTTPRequestHandler class];
+    
+    SEL originalSelector = @selector(URLSession:task:didReceiveChallenge:completionHandler:);
+    SEL swizzledSelector = @selector(authenticationChallenge_URLSession:task:didReceiveChallenge:completionHandler:);
+    
+    Method originalMethod = class_getInstanceMethod(class, originalSelector);
+    Method swizzledMethod = class_getInstanceMethod(class, swizzledSelector);
+    
+    BOOL didAddMethod = class_addMethod(class, originalSelector, method_getImplementation(swizzledMethod), method_getTypeEncoding(swizzledMethod));
+    
+    if (didAddMethod) {
+      class_replaceMethod(class, swizzledSelector, method_getImplementation(originalMethod), method_getTypeEncoding(originalMethod));
+    } else {
+      method_exchangeImplementations(originalMethod, swizzledMethod);
+    }
+  });
+}
+
+/* This method extends the RCTHTTPRequestHandler by intercepting the authentication challenge,
+ * and gives to it a custom behavior regarding allowing or denying connections.
+ * The current implementation of RCTHTTPRequestHandler doesn't provide an implementation for
+ * the authentication challenge event, therefore we don't risk to interfere with React Native behavior.
+ * See the original code:
+ * https://github.com/facebook/react-native/blob/v0.72.5/packages/react-native/Libraries/Network/RCTHTTPRequestHandler.mm
+ *
+ * It worths a clarification regarding the role of RCTHTTPRequestHandler in the context of the
+ * iOS network framework. It plays the role of a NSURLSessionDelegate of the kind NSURLSessionDataDelegate.
+ * A delegate is a callback injected in the NSURLSession, and aims to intercept each phase of a connection.
+ * In React Native, when a session is created, this delegate is injected in.
+ */
+- (void)authenticationChallenge_URLSession:(NSURLSession *)session
+                                      task:(NSURLSessionTask *)task
+                       didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
+                         completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition disposition, NSURLCredential *credential))completionHandler {
+  // Set of allowed domains
+  NSSet *allowedDomains = [self getAllowedDomains];
```

Comment:
Why are we running this for every request? Isn't it immutable? I guess it might have some performance impact. Caching should work fine.

Reply:
You right, done at: e447699556130341b2c10bd3d9b32c90416f34c3